### DISCUSSION TITLE
Plugin system for importing non-RINEX files in ´rinex-cli`

### DISCUSSION THREAD

**Discussion by larsnaesbye** [nav-solutions/rinex#134] [Ideas]
This is an idea that came out of my sparse work of trying to look at writing a Septentrio to RINEX converter.

`teqc` can import non-RINEX files in various formats, e.g. the vendor-specific 'RAW' data files. This functionality comes from code donated to tecq from the vendors but this in return made it impossible to open source license teqc. This seems to have been a major reason for some to hold on to tecq.

So, here's my idea: Expanding `rinex-cli` with the functionality to detect compiled plugins that conform to a simple specification, so the plugins can process conversion from external format to RINEX before other processing.

Upon launch, `rinex-cli` would act on an import flag, check for a plugin that matches and dynamically link to the plugins code and offload the conversion to the plugin, perhaps by passing a memory chunk with the data or just a filepath. The plugin then does its processing, returns a `Rinex` object and is unloaded from memory. `rinex-cli` then does the further processing as before as if it had loaded a RINEX file.

Do you see potential in this or just bloat?

**Comment by gwbres** [comment]
Hello @larsnaesbye,

> This seems to have been a major reason for some to hold on to tecq

definitely

> Do you see potential in this or just bloat?

Yes, but we need an interface to pass data to the third party tool, in the form of binary blob exchanges (data we cannot interpret).
Also the third party tool must output or return a compatible structure. If it's `Rinex` then it's most likely either a rust application, or a python application. Or create a RINEX file itself that we can then parse (this probably exists).

**Comment by larsnaesbye** [comment]
> Yes, but we need an interface to pass data to the third party tool, in the form of binary blob exchanges (data we cannot interpret).

My thought was that we support only one type of plugin - a library with a few defined and exposed functions so rinex-cli can get metadata from it, pass a blob to it and get a RINEX structure back. If the easiest way is to limit it to Rust libraries, that's fine.

> Also the third party tool must output or return a compatible structure. If it's `Rinex` then it's most likely either a rust application, or a python application.

I like the idea of returning a `Rinex` structure, but there are probably other ways too.

**Comment by larsnaesbye** [comment]
I suggest this is something way after v1.0.0, but I would take it into consideration when working on my converter so it could turn into a plugin if we go down that path.
